nh_clinical/activity_extension.py

```diff
@@ -538,110 +530,241 @@ def get_activity_user_ids(self, cr, uid, activity_id, context=None):
+        """
+        Get the create_data dictionary for the policy
+
+        :param case: Case that controls the logic used when getting policy
+        :type case: int
+        :return: Dictionary of values to use when creating a new activity
+        :rtype: dict
+        """
+        return {}
+
+    @staticmethod
+    def trigger_policy_create_activity(source_activity, model, case=None):
+        """
+        Create a new activity using the data defined in the policy with the
+        source activity as it's creator
+
+        :param source_activity: Activity to use to get activity info from
+        :type source_activity: nh.activity record
+        :param model: Model to create activity on
+        :type model: Odoo model
+        :param case: Case to use when getting the data to create the new
+            activity with
+        """
+        new_activity_data = {
+            'patient_id': source_activity.patient_id.id
+        }
+        # Get the create data from the model instance (a subclass of this one)
+        # so that models can define the data themselves
+        new_activity_data.update(
+            source_activity.data_ref._get_policy_create_data(case))
+        return model.sudo(SUPERUSER_ID).create_activity(
+            {
+                'patient_id': source_activity.patient_id.id,
+                'parent_id': source_activity.parent_id.id,
+                'spell_activity_id': source_activity.spell_activity_id.id,
+                'creator_id': source_activity.id
+            },
+            new_activity_data
+        )
+
+    @api.model
+    def get_schedule_date(self, activity, recurring=False):
+        """
+        Get the time to schedule the activity at
+
+        :param activity: Activity to get schedule date for
+        :param recurring: If the schedule date needs to take into consideration
+            recurrance for the activities
+        :return: Date to schedule the activity for
+        :rtype: datetime
+        """
+        if recurring:
+            date_schedule = \
+                self.get_recurring_activity_date_scheduled(activity)
+        else:
+            date_schedule = dt.now() + td(minutes=60)
+        return date_schedule
+
+    @api.model
+    def trigger_policy_activity(
+            self, source_activity, activity_dict, case=None):
+        """
+        Trigger the activity as defined in the policy dictionary
+
+        :param source_activity: Record of the activity that's going to be used
+            as the base for the new activities
+        :type source_activity: nh.activity record
+        :param activity_dict: Activity dictionary contains:
+            - model
+            - type
+            - cancel_others (optional)
+        :type activity_dict: dict
+        :param case: Case to use when triggering policy
+        :type case: int
+        """
+        triggered_model = self.env[activity_dict['model']]
+        if activity_dict.get('cancel_others'):
```

Comment:
Rather pedantic but I expect `if cancel_others in activity_dict:` will be cheaper.

Reply:
I'm not checking that it's in the dict, I'm using the value too